addons/binding/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/api/HeosSystem.java

```diff
@@ -0,0 +1,544 @@
+        groupMapOld.put(hashValue, heosGroup);
+    }
+
+    public List<Map<String, String>> getFavorites() {
+        send(command().browseSource(FAVORIT_SID));
+        return heosDecoder.getPayloadList();
+    }
+
+    public List<String> getPlaylists() {
+        List<String> playlistsList = new ArrayList<>();
+        send(command().browseSource(PLAYLISTS_SID));
+        List<Map<String, String>> payload = heosDecoder.getPayloadList();
+        for (int i = 0; i < payload.size(); i++) {
+            playlistsList.add(payload.get(i).get(CID));
+        }
+        return playlistsList;
+    }
+
+    public HeosFacade getAPI() {
+        return heosApi;
+    }
+
+    public String getConnectionIP() {
+        return connectionIP;
+    }
+
+    public void setConnectionIP(String connectionIP) {
+        this.connectionIP = connectionIP;
+    }
+
+    public int getConnectionPort() {
+        return connectionPort;
+    }
+
+    public void setConnectionPort(int connectionPort) {
+        this.connectionPort = connectionPort;
+    }
+
+    public Map<String, HeosPlayer> getPlayerMap() {
+        return playerMapNew;
+    }
+
+    public Map<String, HeosGroup> getGroupMap() {
+        return groupMapNew;
+    }
+
+    public Map<String, HeosGroup> getGroupsRemoved() {
+        return removedGroupMap;
+    }
+
+    public Map<String, HeosPlayer> getPlayerRemoved() {
+        return removedPlayerMap;
+    }
+
+    /**
+     * A class which provides a runnable for the HEOS Heart Beat
+     *
+     * @author Johannes Einig
+     */
+    public class KeepAliveRunnable implements Runnable {
+        @Override
+        public void run() {
+            try {
+                if (sendCommand.isConnectionAlive()) {
+                    logger.debug("Sending Heos Heart Beat");
+                    if (!sendCommand.send(command().heartbeat())) {
+                        logger.debug("Connection to HEOS Network lost!");
+                        restartConnection();
+                    }
+                } else {
+                    logger.debug("Connection to HEOS Network lost!");
+                    restartConnection();
+                }
+                // catches a failure during a heart beat send message if connection was
+                // getting lost between last Heart Beat but Bridge is online again and not
+                // detected by isConnectionAlive()
+            } catch (ReadException | IOException e) {
+                logger.debug("Failure during HEOS Heart Beat command with message: {}", e.getMessage());
+                restartConnection();
+            }
+        }
+
+        private void restartConnection() {
+            closeConnection();
+            eventController.connectionToSystemLost();
+            try {
+                while (!sendCommand.isConnectionAlive()) {
+                    logger.debug("Trying to reconnect to HEOS Network...");
+                    Thread.sleep(5000);
```

Comment:
Here as well you should use scheduled tasks. WIth the current implementation, you will have a hanging thread if the binding is stopped at that moment. Instead, it should immediately disconnect, end any activity and return from all methods.